client/components/swipeable/index.js

```diff
@@ -0,0 +1,249 @@
+import classnames from 'classnames';
+import { useRtl } from 'i18n-calypso';
+import { Children, useState, useEffect, useRef } from 'react';
+
+import './style.scss';
+
+export const Swipeable = ( {
+	hasDynamicHeight = false,
+	children,
+	currentPage = 0,
+	onPageSelect,
+	pageClassName,
+	containerClassName,
+} ) => {
+	const [ pageWidth, setPageWidth ] = useState( null );
+	const [ swipeableArea, setSwipeableArea ] = useState();
+	const isRtl = useRtl();
+
+	const [ pagesStyle, setPagesStyle ] = useState( {
+		transform: `translate3d(0px, 0px, 0px)`,
+		transitionDuration: `300ms`,
+	} );
+	const [ dragStartData, setDragStartData ] = useState();
+
+	const pagesRef = useRef();
+	const resizeObserverRef = useRef();
+	const numPages = Children.count( children );
+
+	const getWidth = () => {
+		return resizeObserverRef.current?.getBoundingClientRect().width;
+	};
+
+	const getPagesWidth = () => {
+		return pageWidth * numPages;
+	};
+
+	const getOffset = ( index ) => {
+		const offset = pageWidth * index;
+		return isRtl ? offset : -offset;
+	};
+
+	function useUpdateLayout( enabled, currentPageIndex, updateLayout ) {
+		// save callback to a ref so that it doesn't need to be a dependency of other hooks
+		const savedUpdateLayout = useRef();
+		useEffect( () => {
+			savedUpdateLayout.current = updateLayout;
+		}, [ updateLayout ] );
+
+		// fire when the `currentPageIndex` parameter changes
+		useEffect( () => {
+			if ( ! enabled ) {
+				return;
+			}
+
+			savedUpdateLayout.current();
+		}, [ enabled, currentPageIndex ] );
+
+		// fire when the window resizes
+		useEffect( () => {
+			if ( ! enabled ) {
+				return;
```

Comment:
Wouldn't this early return prevent the event listeners from being cleared? We should probably do instead:

```
if ( enabled ) {
  // add ev listeners
}

return {
  if (enabled) {
    // remove event listeners
  }
}

Reply:
Can you explain a bit more when this would be the case?

Reply:
Nevermind my initial comment, I misread the code.

The `useEffect` hook returns a [cleanup function](https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect) — I initially thought that, by having an early return, we would miss running the cleanup callback (the one that removes the `resize` and `orientationchange` ev listeners).

But actually, by returning early, we also don't add any of these ev. listeners — and therefore there's nothing to cleanup.

Going to mark this as resolved, sorry!